dm/tracer/store.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"sync"
+
+	"github.com/pingcap/errors"
+
+	"github.com/pingcap/dm/dm/pb"
+)
+
+// TraceEvent represents a sigle tracing event
+type TraceEvent struct {
+	Type  pb.TraceType `json:"type"`
+	Event interface{}  `json:"event"`
+}
+
+// EventStore stores all tracing events, mapping from TraceID -> a list of TraceEvent
+// TraceEvents with the same TraceID can have different TraceType
+// NOTE: this is a quick and dirty implement, we will refactor it later
+type EventStore struct {
+	sync.RWMutex
+	ids    []string
+	events map[string][]*TraceEvent
+}
+
+// NewEventStore creates a new EventStore
+func NewEventStore() *EventStore {
+	return &EventStore{
+		events: make(map[string][]*TraceEvent),
+	}
+}
+
+func (store *EventStore) addNewEvent(e *TraceEvent) error {
+	var traceID string
+	switch e.Type {
+	case pb.TraceType_BinlogEvent:
+		ev, ok := e.Event.(*pb.SyncerBinlogEvent)
+		if !ok {
+			return errors.NotValidf("trace event data")
+		}
+		traceID = ev.Base.TraceID
+	case pb.TraceType_JobEvent:
+		ev, ok := e.Event.(*pb.SyncerJobEvent)
+		if !ok {
+			return errors.NotValidf("trace event data")
+		}
+		traceID = ev.Base.TraceID
+	default:
+		return errors.NotValidf("trace event type %d", e.Type)
+	}
+
+	store.Lock()
+	_, ok := store.events[traceID]
+	if !ok {
+		store.ids = append(store.ids, traceID)
+		store.events[traceID] = make([]*TraceEvent, 0)
+	}
+	store.events[traceID] = append(store.events[traceID], e)
+	store.Unlock()
+
+	return nil
+}
+
+// queryByTraceID queries trace event by given traceID
+func (store *EventStore) queryByTraceID(traceID string) []*TraceEvent {
+	store.RLock()
+	defer store.RUnlock()
+	return store.events[traceID]
+}
+
+func (store *EventStore) scan(offset, limit int64) [][]*TraceEvent {
+	store.RLock()
+	defer store.RUnlock()
+	if offset > int64(len(store.ids)) {
+		return nil
+	}
+	result := make([][]*TraceEvent, 0, limit)
+	for idx := offset; idx < offset+limit; idx++ {
+		if idx >= int64(len(store.ids)) {
```

Comment:
you can pre-calculate the `end` outer of the `for` loop. minimus of `offset+limit` and `len(store.ids)`?